### PR TITLE
http: proxying 102s.

### PR DESCRIPTION
**Pull request by alyssawilk**
Proxying 102s from upstream

Risk Level: Low (minor refactor)
Testing: new integration test
Docs Changes: n/a
Release Notes: inline
Runtime guard: envoy.reloadable_features.proxy_102
Fixes https://github.com/envoyproxy/envoy/issues/18844

### CODE REVIEW

**Comment by alyssawilk** [comment]
cc @RyanTheOptimist @DavidSchinazi now with 103s :-)

File: test/integration/protocol_integration_test.cc [hunk 1]
```diff
@@ -1341,6 +1341,10 @@ TEST_P(ProtocolIntegrationTest, EnvoyProxyingLateMultiple1xx) {
   testEnvoyProxying1xx(false, false, true);
 }
 
+TEST_P(ProtocolIntegrationTest, EnvoyProxying102) {
```
Comment (RyanTheOptimist) [inline test/integration/protocol_integration_test.cc:1344]: Should we also have a 103 test here?

**Review by RyanTheOptimist — DISMISSED** [review]
Woo hoo! Looks awesome. One small question.

File: docs/root/version_history/current.rst [hunk 2]
```diff
@@ -15,6 +15,7 @@ Minor Behavior Changes
 * bandwidth_limit: added :ref:`bandwidth limit stats <config_http_filters_bandwidth_limit>` *request_enforced* and *response_enforced*.
 * config: the log message for "gRPC config stream closed" now uses the most recent error message, and reports seconds instead of milliseconds for how long the most recent status has been received.
 * dns: now respecting the returned DNS TTL for resolved hosts, rather than always relying on the hard-coded :ref:`dns_refresh_rate. <envoy_v3_api_field_config.cluster.v3.Cluster.dns_refresh_rate>` This behavior can be temporarily reverted by setting the runtime guard ``envoy.reloadable_features.use_dns_ttl`` to false.
+* http: envoy will now proxy 102 and 103 headers from upstream, though as with 100s only the first 1xx response headers will be sent. This behavioral change by can temporarily revertd by setting runtime guard ``envoy.reloadable_features.proxy_102_103`` to false.
```
Comment (DavidSchinazi) [inline docs/root/version_history/current.rst:18]:
```suggestion
* http: envoy will now proxy 102 and 103 headers from upstream, though as with 100s only the first 1xx response headers will be sent. This behavioral change by can temporarily reverted by setting runtime guard ``envoy.reloadable_features.proxy_102_103`` to false.
```

**Review by DavidSchinazi — APPROVED** [review]

**Review by snowp — APPROVED** [review]
LGTM, thanks!

**Pull request merged**